lib/ls.js

```diff
@@ -117,12 +117,18 @@ function filterByEnv (data) {
   var dev = npm.config.get('dev') || /^dev(elopment)?$/.test(npm.config.get('only'))
   var production = npm.config.get('production') || /^prod(uction)?$/.test(npm.config.get('only'))
   var dependencies = {}
-  var devDependencies = data.devDependencies || []
+  var devKeys = Object.keys(data.devDependencies || [])
+  var prodKeys = Object.keys(data._dependencies || [])
   Object.keys(data.dependencies).forEach(function (name) {
-    var keys = Object.keys(devDependencies)
-    if (dev && !production && keys.indexOf(name) === -1) return
-    if (!dev && keys.indexOf(name) !== -1 && data.dependencies[name].missing) return
-    dependencies[name] = data.dependencies[name]
+    if (!dev && devKeys.indexOf(name) !== -1 && data.dependencies[name].missing) {
```

Comment:
This uses `!== -1` where as later you use `>= 0`– I'll standardize these when I merge this.